### PR TITLE
Assemblers for 8-bit chips

### PR DESCRIPTION
**Pull request by AndersonTorres**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 137142` at 4f1c9406 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built successfully:</summary>
  <ul>
    <li>as31</li>
    <li>atasm</li>
  </ul>
</details>

File: pkgs/development/compilers/as31/default.nix [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+  patches = [
+    # Check return value of getline in run.c
+    ./0000-getline-break.patch
+  ];
+
+  postPatch = ''
+    # parser.c is generated from parser.y; it is better to generate it via bison
+    # instead of using the prebuilt one, especially in x86_64
+    rm -f as31/parser.c
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/as31/default.nix:25]: The -f should be removed.

File: pkgs/development/compilers/atasm/default.nix [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  nativeBuildInputs = [
+    unzip
+  ];
+
+  buildInputs = [
+    zlib
+  ];
+
+  preBuild = ''
+    makeFlagsArray+=(
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/atasm/default.nix:35]: Use makeFlags instead.

File: pkgs/development/compilers/atasm/default.nix [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+      CC=cc
+      USEZ="-DZLIB_CAPABLE -I${zlib}/include"
+      ZLIB="-L${zlib}/lib -lz"
+      UNIX="-DUNIX"
+    )
+  '';
+
+  preInstall = ''
+    install -d $out/share/doc/${pname} $out/man/man1
+    installFlagsArray+=(
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/atasm/default.nix:46]: installFlags

File: pkgs/development/compilers/atasm/default.nix [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+      -C ./src
+      CC=cc
+      USEZ="-DZLIB_CAPABLE -I${zlib}/include"
+      ZLIB="-L${zlib}/lib -lz"
+      UNIX="-DUNIX"
+    )
+  '';
+
+  preInstall = ''
+    install -d $out/share/doc/${pname} $out/man/man1
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/atasm/default.nix:45]: This should use installManPage.

File: pkgs/development/compilers/atasm/default.nix (hunk 2 above)
Comment (AndersonTorres) [reply pkgs/development/compilers/atasm/default.nix:35]: What about the whitespaces?

Comment (SuperSandro2000) [reply pkgs/development/compilers/atasm/default.nix:35]: What do you mean? The quoting in makeFlags is not suprt important and in the end it is bash things anyway.

Comment (AndersonTorres) [reply pkgs/development/compilers/atasm/default.nix:35]: How to write this in `makeFlags`?

`USEZ="-DZLIB_CAPABLE -I${zlib}/include"`

Comment (SuperSandro2000) [reply pkgs/development/compilers/atasm/default.nix:35]:
```nix
makeFlags = [ "USEZ=\"-DZLIB_CAPABLE -I${zlib}/include\"" ]
```
or
```nix
makeFlags = [ ''USEZ="-DZLIB_CAPABLE -I${zlib}/include"'' ]
```

Comment (AndersonTorres) [reply pkgs/development/compilers/atasm/default.nix:35]:
```
       > build flags: SHELL=/nix/store/dpjnjrqbgbm8a5wvi1hya01vd8wyvsq4-bash-4.4-p23/bin/bash -C ./src CC=cc USEZ=\"-DZLIB_CAPABLE -I/nix/store/dc3r2nbhckwn41hq5vaxxbj7z0lqdwb1-zlib-1.2.11/include\" ZLIB=\"-L/nix/store/dc3r2nbhckwn41hq5vaxxbj7z0lqdwb1-zlib-1.2.11/lib -lz\" UNIX=\"-DUNIX\"
       > make: Entering directory '/build/atasm109/src'
       > cc -g -Wall "-DZLIB_CAPABLE  "-DUNIX"  -c asm.c
       > /nix/store/dpjnjrqbgbm8a5wvi1hya01vd8wyvsq4-bash-4.4-p23/bin/bash: -c: line 0: unexpected EOF while looking for matching `"'
       > /nix/store/dpjnjrqbgbm8a5wvi1hya01vd8wyvsq4-bash-4.4-p23/bin/bash: -c: line 1: syntax error: unexpected end of file
       > make: *** [Makefile:29: asm.o] Error 1
       > make: Leaving directory '/build/atasm109/src'
       For full logs, run 'nix log /nix/store/az0zws725h9rk6paagw9dshxcbwjsr69-atasm-1.09.drv'.
```

File: pkgs/development/compilers/atasm/default.nix (hunk 4 above)
Comment (AndersonTorres) [reply pkgs/development/compilers/atasm/default.nix:45]: Nope. The Makefile can install the manpage, however it does require a pre-existent directory.

**Pull request merged**